File: LeetCode/104_MaximumDepthOfBinaryTree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var maxDepth = function(root) {
  if (!root) {
    return 0;
  }

  return Math.max.apply(null, [root.left, root.right].map(function(child) {
    return child ? 1 + maxDepth(child) : 1;
  }));
};

File: LeetCode/lib/104_MaximumDepthOfBinaryTree.rb
# Given a binary tree, find its maximum depth.
#
# The maximum depth is the number of nodes along the longest path from the root
# node down to the farthest leaf node.
#
# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val)
#         @val = val
#         @left, @right = nil, nil
#     end
# end

# @param {TreeNode} root
# @return {Integer}
def max_depth(root)
  return 0 unless root

  [root.left, root.right].map do |child|
    child ? 1 + max_depth(child) : 1
  end.max
end
